### PR TITLE
fix: username generation for new user and modify format format func for migration

### LINKED ISSUE #2133: Fix Username Generation for New Users During Signup Based on Migration Logic
### Description:
The current username generation process during user signup does not comply with the new migration logic introduced for existing users. This issue aims to update the username generation for new users to ensure consistency with the migration logic, which enforces a specific format, length constraint, and conflict resolution for duplicate usernames.

### Acceptance Criteria:
**Username Format:**

The username should follow the format: {firstname}-{lastname}-{n}.
First name: The first part of the first name (ignore middle names).
Last name: Truncated as needed to fit within the 32-character limit.
Suffix: A numeric suffix starting from -1 for the first occurrence of the name and incremented for subsequent users with the same name combination.

**Character Limit:**

The total username length should not exceed 32 characters, including the first name, last name, hyphens, and suffix.
If the total length exceeds the limit, only the last name should be truncated to fit within the character limit.

**Edge Case Handling:**

Middle Names: If a user’s first name includes a middle name, only the first part (e.g., Vinit from Vinit Kumar) should be used.
Whitespace: Ensure that leading and trailing whitespaces are removed from both the first name and last name during processing.
Conflict Resolution: When a username conflict arises (i.e., if a username already exists), append a numeric suffix starting with -1 and increment for each subsequent user with the same name combination.
Example: If the first user with the name John Doe has the username john-doe-1, the next user with the same name should be assigned john-doe-2.

### PR DESCRIPTION
**Pull request by vinit717**
## Issue Ticket Number
- closes #2133

## Description

In this PR username generation functionality is modified and make format func is generic so It can be used in migration
 https://docs.google.com/document/d/1qO9QqBR6cgmFzC4UEnehSVugAOThtY0u90BsHFqrcUw/edit?usp=sharing

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
image

## Test Coverage
image
image
image

### CODE REVIEW

File: models/users.js [hunk 1]
```diff
@@ -790,21 +791,17 @@ const getUsersByRole = async (role) => {
...
-    const initialUsername = `${firstname}-${lastname}`;
-    if (existingUserCount === 0) {
-      return initialUsername;
-    } else {
-      const uniqueUsername = `${initialUsername}-${existingUserCount + 1}`;
-      return uniqueUsername;
-    }
+    const snapshot = await userModel.where("first_name", "==", firstName).where("last_name", "==", lastName).get();
+
+    const existingUserCount = snapshot.size;
```
Comment (prakashchoudhary07) [inline models/users.js:798]: I think there is something in firestore that directly returns count have a look once

**Comment by vinit717** [comment]
#2147

File: test/unit/models/users.test.js [hunk 2]
```diff
@@ -425,6 +425,26 @@ describe("users", function () {
...
+
+    it("should generate a unique username when no existing users are present", async function () {
+      const firstName = "john";
+      const lastName = "doe";
+
+      const whereStub = sinon.stub(userModel, "where").returnsThis();
+      const countStub = sinon.stub().returnsThis();
+      const getStub = sinon.stub().resolves({
+        data: () => ({ count: 0 }),
+      });
```
Comment (prakashchoudhary07) [inline test/unit/models/users.test.js:437]: Why are we mocking this?
We are using firestore emulator right

File: utils/username.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    const validFirstName = /^[a-zA-Z]+$/.test(firstNamePart)
+        ? firstNamePart.toLowerCase()
+        : "null";
+
+    const lastNameParts = trimmedLastName.split(/[\s-]/);
+    const lastNamePart = lastNameParts[lastNameParts.length - 1];
+    const validLastName = /^[a-zA-Z]+$/.test(lastNamePart)
+        ? lastNamePart.toLowerCase()
+        : "null";
+
```
Comment (prakashchoudhary07) [inline utils/username.ts:17]: What are the regex for?

Comment (vinit717) [reply utils/username.ts:17]: To validate the characters
for more, you can look format username tests

File: test/unit/models/users.test.js (hunk 2 above)
Comment (vinit717) [reply test/unit/models/users.test.js]: fixed

File: models/users.js (hunk 1 above)
Comment (vinit717) [reply models/users.js]: fixed

File: models/users.js [hunk 4]
```diff
@@ -790,21 +791,20 @@ const getUsersByRole = async (role) => {
   }
 };
 
-const generateUniqueUsername = async (firstname, lastname) => {
+const generateUniqueUsername = async (firstName, lastName) => {
```
Comment (iamitprakash) [inline models/users.js:794]:
```
const generateUniqueUsername = async (firstName, lastName) => {
  try {
    let suffix = 1;
    let finalUsername;
    let isUnique = false;

    while (!isUnique) {
      finalUsername = formatUsername(firstName, lastName, suffix);
      
      const snapshot = await userModel
        .where("username", "==", finalUsername)
        .count()
        .get();
      
      const existingCount = snapshot.data().count || 0;
      
      if (existingCount === 0) {
        isUnique = true;
      } else {
        suffix++;
      }
    }

    return finalUsername;
  } catch (err) {
    logger.error(`Error while generating unique username: ${err.message}`);
    throw err;
  }
};
```

Comment (prakashchoudhary07) [inline models/users.js:794]: Please move this out of model

Comment (vinit717) [reply models/users.js]: Done

File: middlewares/validators/user.js [hunk 5]
```diff
@@ -328,6 +338,7 @@ const validateGenerateUsernameQuery = async (req, res, next) => {
     res.boom.badRequest("Invalid Query Parameters Passed");
   }
 };
```
Comment (iamitprakash) [inline middlewares/validators/user.js:340]: please remove white space

**Review by prakashchoudhary07 — APPROVED** [review]
LGTM

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**